hal/src/sercom/v2/i2c.rs

```diff
@@ -0,0 +1,431 @@
+//!   slices, consider using the DMA methods instead. The `dma` Cargo feature
+//!   must be enabled.
+//!
+//! [`enable`]: Config::enable
+//! [`disable`]: I2c::disable
+//! [`reconfigure`]: I2c::reconfigure
+//! [`bsp_pins`]: crate::bsp_pins
+//! [`Sercom`]: crate::sercom::v2::Sercom
+//! [`Pad0`]: crate::sercom::v2::pad::Pad0
+//! [`Pad1`]: crate::sercom::v2::pad::Pad1
+//! [`Pad`]: crate::sercom::v2::pad::Pad
+//! [`IsPad`]: crate::sercom::v2::pad::IsPad
+//! [`PadNum`]: crate::sercom::v2::pad::PadNum
+//! [`v1::Pin`]: crate::gpio::v1::Pin
+//! [`v2::Pin`]: crate::gpio::v2::pin::Pin
+//! [`Pin`]: crate::gpio::v2::pin::Pin
+//! [`PinId`]: crate::gpio::v2::pin::PinId
+//! [`PinMode`]: crate::gpio::v2::pin::PinMode
+//! [`i2c::Write`]: embedded_hal::blocking::i2c::Write
+//! [`i2c::Read`]: embedded_hal::blocking::i2c::Read
+//! [`i2c::WriteRead`]: embedded_hal::blocking::i2c::WriteRead
+#![cfg_attr(
+    feature = "dma",
+    doc = "
+# Using I2C with DMA
+
+This HAL includes support for DMA-enabled I2C transfers. [`I2c`]
+implements the DMAC [`Buffer`]
+trait. The provided [`send_with_dma`] and
+[`receive_with_dma`] build and begin a
+[`dmac::Transfer`], thus starting the I2C
+in a non-blocking way. 
+
+Note that the [`init_dma_transfer`] method should be called immediately before
+starting a DMA transfer with I2C. This will check that the bus is in a correct
+state before starting the transfer, and providing a token type to pass to the 
+[`send_with_dma`] and [`receive_with_dma`] methods.
+
+Optionally, interrupts can be enabled on the provided
+[`Channel`]. Note that the `dma` feature must
+be enabled. Please refer to the [`dmac`](crate::dmac) module-level
+documentation for more information.
+
+```no_run
+use atsamd_hal::dmac::channel::{AnyChannel, Ready};
+use atsand_hal::sercom::v2::i2c::{I2c, AnyConfig, Error};
+fn i2c_send_with_dma<A: AnyConfig, C: AnyChannel<Status = Ready>>(i2c: I2c<A>, channel: C) -> Result<(), Error>{
+    // Create data to send. Note that it must be `'static`.
+    let buffer: [u8; 50] = [0xff; 50];
+
+    // Initialize the bus and check for errors
+    let token = i2c.init_dma_transfer()?;
+    let transfer = i2c.send_with_dma(0x54, token, buffer, channel0, |_| {})
+
+
+    // Wait for transfers to complete and reclaim resources
+    let (chan0, buffer, i2c) = transfer.wait();
+
+    // Check for errors that may have occured during the transfer.
+    i2c.read_status().try_into()?;
+}
+```
+
+[`Buffer`]: crate::dmac::transfer::Buffer
+[`init_dma_transfer`]: I2c::init_dma_transfer
+[`send_with_dma`]: I2c::send_with_dma
+[`receive_with_dma`]: I2c::receive_with_dma
+[`dmac::Transfer`]: crate::dmac::Transfer
+[`Channel`]: crate::dmac::channel::Channel
+[`dmac`]: crate::dmac
+
+"
+)]
+
+#[cfg(any(feature = "samd11", feature = "samd21"))]
+#[path = "i2c/pads_thumbv6m.rs"]
+mod pads;
+
+#[cfg(feature = "min-samd51g")]
+#[path = "i2c/pads_thumbv7em.rs"]
+mod pads;
```

Comment:
I realized that the `path` specifier may not be necessary. Wouldn't this do the same thing?

```rust
mod pads_thumbv7em;

use pads_thumbv7em as pads;

pub use pads::*;
```

Reply:
It is, but it takes a `#[cfg]` for both the `mod` and the `use` statements. I guess it's a matter of preference, but to me the single `#[cfg]` version seems a little less error-prone.